### PR TITLE
py-tensorflow: Update to version 2.12.0, Add Python 310

### PR DESCRIPTION
**Pull request by essandess**
#### Description

* Depends upon #16149.

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 12.4 21F79 x86_64
Xcode 13.4.1 13F100

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: devel/bazel/Portfile
Comment (mascguy) [inline devel/bazel/Portfile:76]: Based on a quick lint check, it looks like you have trailing whitespace on line 76.

Comment (essandess) [reply devel/bazel/Portfile]: Thanks, fixed.

**Comment by essandess** [comment]
> Steve, I ran a quick local test on 10.15, to help validate whether the CI failures are legitimate. And sure enough, `bazel` is failing to build for me as well.

I only have access to macOS 12, and the build works. I expect that's an upstream issue that we'll never have much control over, or cycles to fix (per https://lists.macports.org/pipermail/macports-dev/2022-July/044443.html).

I suggest just getting this thing running on the latest macOS and call it a day. If anyone requires bazel on older macOS version and knows how to fix it, or wants to download the binaries directly, they can issue a PR for an updated approach.

**Comment by essandess** [comment]
[Upstream](https://github.com/bazelbuild/bazel/issues/15876#issuecomment-1196959184) offers three options to address old bazel version issues:
* Figure out a patch ourselves
* Use a tool called `bazelisk` to download and install previous binaries
* Get any repos that use prior bazel versions to support bazel 5

None of these are viable options at least with the resources I have. This supports adoption of the strategy laid out above:

> just get this thing running on the latest macOS and call it a day. If anyone requires bazel on older macOS version and knows how to fix it, or wants to download the binaries directly, they can issue a PR for an updated approach.

File: python/py-tensorflow/Portfile
Comment (mascguy) [inline python/py-tensorflow/Portfile:65]: Given that `py-tensorflow_macos` is obsolete, let's remove that conflict declaration.

Comment (mascguy) [reply python/py-tensorflow/Portfile]: > Given that `py-tensorflow_macos` is obsolete, let's remove that conflict declaration.

Done. Also deleted that port, along with removing all references to it.

Comment (mascguy) [reply python/py-tensorflow/Portfile]: > > Given that `py-tensorflow_macos` is obsolete, let's remove that conflict declaration.
> 
> Done. Also deleted that port, along with removing all references to it.

No need to wait for this PR, so I took of deletion on master. Also rebased to pick up those changes.

**Comment by essandess** [comment]
This now fails locally with the linker error:
```sh
:info:build # Execution platform: @local_execution_config_platform//:platform
:info:build ld: malformed trie, childNodeOffset==0 file 'bazel-out/darwin-opt/bin/_solib_darwin_x86_64/_U_S_Stensorflow_Spython_C_Upywrap_Utensorflow_Uinternal_Umacos___Utensorflow_Spython/lib_pywrap_tensorflow_internal.dylib'
:info:build clang: error: linker command failed with exit code 1 (use -v to see invocation)
:info:build Error in child process '/usr/bin/xcrun'. 1
```

I'm throwing in the towel getting tensorflow built, and will split off the `bazel` commits, which do build and run.

**Comment by reneeotten** [comment]
@essandess so should this PR be closed or is there still work being done to get this to a merge-able state?

**Comment by essandess** [comment]
> @essandess so should this PR be closed or is there still work being done to get this to a merge-able state?

I would defer to the port maintainers.

I don’t know how to build either the latest versions of `tensorflow` or `bazel`, and suspect the only stable approach now is to simply download prebuilt binaries.

**Comment by essandess** [comment]
FWIW, I observe that TF 2.11.0 fails to build with the latest `bazel`:
```
:info:build ERROR: /opt/local/var/macports/build/_opt_local_ports_python_py-tensorflow/py310-tensorflow/work/tensorflow-tensorflow-d5b57ca/tensorflow/python/BUILD:358:27: Linking tensorflow/python/_pywrap_py_exception_registry.so failed: (Aborted): cc_wrapper.sh failed: error executing command
:info:build   (cd /opt/local/var/macports/build/_opt_local_ports_python_py-tensorflow/py310-tensorflow/work/bazel_build/4031e1b921605e62c062b5347ee6e130/execroot/org_tensorflow && \
:info:build   exec env - \
:info:build     APPLE_SDK_PLATFORM=MacOSX \
:info:build     APPLE_SDK_VERSION_OVERRIDE=13.0 \
:info:build     PATH=/opt/local/var/macports/build/_opt_local_ports_python_py-tensorflow/py310-tensorflow/work/bazelwrap:/opt/local/bin:/opt/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin \
:info:build     PYTHON_BIN_PATH=/opt/local/bin/python3.10 \
:info:build     PYTHON_LIB_PATH=/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages \
:info:build     TF2_BEHAVIOR=1 \
:info:build     XCODE_VERSION_OVERRIDE=14.1.0.14B47b \
:info:build     ZERO_AR_DATE=1 \
:info:build   external/local_config_cc/cc_wrapper.sh @bazel-out/darwin-opt/bin/tensorflow/python/_pywrap_py_exception_registry.so-2.params)
:info:build # Configuration: cadaf1a962a2ec23394d0407fafd53f6fc6ed02f4709fdb0b472915a8f2d61dd
:info:build # Execution platform: @local_execution_config_platform//:platform
:info:build ld: malformed trie, childNodeOffset==0 file 'bazel-out/darwin-opt/bin/_solib_darwin_x86_64/libtensorflow_Spython_S_Upywrap_Utensorflow_Uinternal.so'
:info:build clang: error: linker command failed with exit code 1 (use -v to see invocation)
:info:build Error in child process '/usr/bin/xcrun'. 1
```

**Comment by essandess** [comment]
Tensorflow version 2.12.0 still doesn't build on macOS because of https://github.com/tensorflow/tensorflow/issues/58805.

Nevertheless, this is the same level of brokenness as the existing TF; therefore, I recommend merging so that we can at least have an existing Python 310 port.

**Comment by catap** [comment]
@essandess shall this PR also be rebased? :)

**Comment by mascguy** [comment]
> @essandess shall this PR also be rebased? :)

Just rebased against master, after merging #16149. Let's see how things look, now that Bazel is fixed.

**Comment by catap** [comment]
It fails as:
```
2023-05-03T21:23:49.8697360Z Analyzing: target //tensorflow/tools/pip_package:build_pip_package (581 packages loaded, 24492 targets configured)
2023-05-03T21:23:51.2965950Z Analyzing: target //tensorflow/tools/pip_package:build_pip_package (582 packages loaded, 26874 targets configured)
2023-05-03T21:23:52.8944720Z Analyzing: target //tensorflow/tools/pip_package:build_pip_package (585 packages loaded, 33560 targets configured)
2023-05-03T21:23:53.3244360Z ERROR: /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-tensorflow/py310-tensorflow/work/tensorflow-2.12.0/tensorflow/compiler/xla/pjrt/BUILD:469:11: in cc_library rule //tensorflow/compiler/xla/pjrt:pjrt_future: target '@tf_runtime//:support' is not visible from target '//tensorflow/compiler/xla/pjrt:pjrt_future'. Check the visibility declaration of the former target if you think the dependency is legitimate
2023-05-03T21:23:53.3252760Z ERROR: /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-tensorflow/py310-tensorflow/work/tensorflow-2.12.0/tensorflow/compiler/xla/pjrt/BUILD:469:11: Analysis of target '//tensorflow/compiler/xla/pjrt:pjrt_future' failed
2023-05-03T21:23:53.5959440Z ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: 
2023-05-03T21:23:53.6210930Z INFO: Elapsed time: 135.543s
```

**Comment by mascguy** [comment]
```
2023-05-03T21:23:53.3244360Z ERROR: py310-tensorflow/work/tensorflow-2.12.0/tensorflow/compiler/xla/pjrt/BUILD:469:11:
    in cc_library rule //tensorflow/compiler/xla/pjrt:pjrt_future: target '@tf_runtime//:support'
    is not visible from target '//tensorflow/compiler/xla/pjrt:pjrt_future'.
    Check the visibility declaration of the former target if you think the dependency is legitimate
2023-05-03T21:23:53.3252760Z ERROR: py310-tensorflow/work/tensorflow-2.12.0/tensorflow/compiler/xla/pjrt/BUILD:469:11:
    Analysis of target '//tensorflow/compiler/xla/pjrt:pjrt_future' failed
```

I believe this is the issue Steve mentioned in his earlier comment, from March 30th. But this upstream comment suggests that it might (?) be patchable:
* https://github.com/tensorflow/tensorflow/issues/58825#issuecomment-1346966590

**Comment by catap** [comment]
@mascguy it will be killed by timeout. It is going about 5h30 and doesn't finished to build py310 subport. Right now it's rebuilding whole LLVM. I'm supprising that it did so...

**Comment by catap** [comment]
@mascguy the whole build requires something like 7 hours. We have 3 subports => ~21 hours. See: https://github.com/tensorflow/io/issues/1236

I may add a trick from https://github.com/tensorflow/io/pull/1294 which should decrease time for an hour per build.

Do you agree?

**Comment by mascguy** [comment]
> @mascguy the whole build requires something like 7 hours. We have 3 subports => ~21 hours. See: https://github.com/tensorflow/io/issues/1236
> 
> I may add a trick from https://github.com/tensorflow/io/pull/1294 which should decrease time for an hour per build.
> 
> Do you agree?

Looks promising!

**Comment by catap** [comment]
I'm wondering: won't patching sources broke the cache...

**Comment by catap** [comment]
Here the branch: https://github.com/macports/macports-ports/compare/master...catap:macports-ports:py-tensorflow

Moves much faster.

**Comment by catap** [comment]
Seems that caching doesn't work because we're patching sources :(

**Comment by catap** [comment]
Long story short: I have no idea how to build it in CI time.

**Comment by mascguy** [comment]
> Long story short: I have no idea how to build it in CI time.
> 
> Using prebuild artifacts won't work because we're building it to be installed under `/opt/local` and simple no such artifacts exists.

I'm currently trying to build `py310-tensorflow` locally, via 10.15. And one thing we can do to improve the build time, is to set `bazel.limit_build_jobs no`. Without that disabled, the bazel PG uses only 50% of the available CPUs, as that used to be necessary for our buildbots. (Now that Ryan has upped the memory on all of them to 16gb, it may not be needed anymore.)

Anyhow, uncomment that line, and re-push. Hopefully that should improve the CI time. (And perhaps we can use that when we ultimately commit to master.)

**Comment by catap** [comment]
@mascguy I think that you may remove remote caching because it won't work. I haven't found any way to warm up such cache yet :(

**Comment by catap** [comment]
One possible approach is https://github.com/tsawada/bazel-github-actions-cache but I not sure that it is good idea

**Comment by catap** [comment]
Another approach (much cleaner) use [webdavipfs](https://github.com/markg85/webdavipfs) as free storage for bazel cache :)

**Comment by catap** [comment]
@mascguy may I ask you to remove `remote_cache` variant? It won't work.

**Comment by catap** [comment]
> > Long story short: I have no idea how to build it in CI time.
> 
> At least for testing in your personal repo, you may want to drop support for Python 3.7 and 3.8. That's something we may also want to consider for the port in-general too: Our opt-in stats don't show any installs for those versions within the past 365 days, so there may not be a need to support those old versions:
> 
> * https://ports.macports.org/port/py37-tensorflow/stats/?days=365&days_ago=0
> * https://ports.macports.org/port/py38-tensorflow/stats/?days=365&days_ago=0

Unfortentnly it still requires for about 7h to be build on github CI :(

**Comment by catap** [comment]
@mascguy FYI: I'm making a few more attempts to build it on github CI. If succeed, I'll notify here in couple of hours.

**Comment by mascguy** [comment]
> @mascguy may I ask you to remove `remote_cache` variant? It won't work.

Haven't forgotten about this, just had a lot of other work to commit earlier. I'll try to tackle this tonight though!

**Comment by catap** [comment]
> > @mascguy may I ask you to remove `remote_cache` variant? It won't work.
> 
> Haven't forgotten about this, just had a lot of other work to commit earlier. I'll try to tackle this tonight though!

I'm testing it right now :) Will ping you soon.

**Comment by catap** [comment]
@mascguy the branch: https://github.com/catap/macports-ports/tree/py-tensorflow

If you remove `py310` subport it probably will pass CI time.

I've tried much more agressive build, like:
```
    bazel.extra_build_opts-append \
                            "--jobs=HOST_CPUS*2.0" \
                            "--local_cpu_resources=HOST_CPUS*2.0" \
                            "--local_ram_resources=HOST_RAM*0.8"
```

but it also doesn't help.

Same for cache server. It is possible to enable but we need to warm it up somehow that seems out of scope for that PR.

Thus, the cause of time requirements that it builds almost everything from scratch. Includes ICU, SSL, LLVM, and everything which includes zlib.

**Comment by reneeotten** [comment]
so is the issue here only that there is a timeout on the CI but you guys have verified that it works locally (and should also work the buildbots)?
In the latter case this should just be merged. But admittedly I haven't followed the while discussion here...

**Review by reneeotten — CHANGES_REQUESTED** [review]

**Comment by catap** [comment]
@reneeotten may I ask you to force push to that branch https://github.com/catap/macports-ports/commit/38d2d59a474c1f65d2ff285c0abefc6cb597fc1a which contains only @essandess changes plus my fix to make it works.

I've rejected all speed optimization by making build more agressive, and removing python versions.

I think that it is the best and cleaner changes that can be merged to master to run via buildbot to see how it goes.

**Comment by reneeotten** [comment]
@catap please make the requested changes in this PR.

**Comment by catap** [comment]
@reneeotten I haven’t got access to make a force push into PRs :(

**Comment by mascguy** [comment]
> @reneeotten may I ask you to force push to that branch [catap@38d2d59](https://github.com/catap/macports-ports/commit/38d2d59a474c1f65d2ff285c0abefc6cb597fc1a) which contains only @essandess changes plus my fix to make it works.
> 
> I've rejected all speed optimization by making build more agressive, and removing python versions.
> 
> I think that it is the best and cleaner changes that can be merged to master to run via buildbot to see how it goes.

Kirill, I'll be working on MacPorts work for most of the days, and am happy to help. So do you want me to update this port based on your most recent commit?

**Comment by catap** [comment]
Yes, please.

I’ve run out of ideas how to pass CI here and suggest to keep things simpler, as I proposed at the last version of my branch: simple add an option which fix the build

File: python/py-tensorflow/Portfile
Comment (catap) [reply python/py-tensorflow/Portfile:48]: FYI: IBM’s version of tensorflow builds about 20 hours (!): https://powerci.osuosl.org/job/TensorFlow2_PPC64LE_CPU_Release_Build/

2B08CCA3-65D7-42E6-BB19-6BCA5B9BB179

**Comment by mascguy** [comment]
> Yes, please.
> 
> I’ve run out of ideas how to pass CI here and suggest to keep things simpler, as I proposed at the last version of my branch: simple add an option which fix the build

Done, though I also kept my later update - which drops Python 3.7/3.8, relaxes Clang/Xcode restrictions, and cleans up conflict declarations - as well.

**Comment by reneeotten** [comment]
I'm fine with removing the older Python versions, but then add at least Python 3.11 which is MacPort's default

**Comment by mascguy** [comment]
> I'm fine with removing the older Python versions, but then add at least Python 3.11 which is MacPort's default

That's going to require adding Python 3.11 support to a number of dependencies, which don't yet exist. So I don't think that's feasible for this update:

```
$ port lint --nitpick py311-tensorflow
--->  Verifying Portfile for py311-tensorflow
Error: Unknown dependency: py311-absl
Error: Unknown dependency: py311-astunparse
Error: Unknown dependency: py311-google-pasta
Error: Unknown dependency: py311-grpcio
Error: Unknown dependency: py311-keras
Error: Unknown dependency: py311-opt_einsum
Error: Unknown dependency: py311-protobuf3
Error: Unknown dependency: py311-tensorboard
Error: Unknown dependency: py311-tensorflow_estimator
```

**Comment by mascguy** [comment]
Okay, I'm finished with my cleanup, and I think we've done as much pre-testing as we can. Let's merge, and let the buildbots take on this update.

**Comment by mascguy** [comment]
The builds are failing very early in the process, for Monterey and Ventura Intel. (Still waiting to see whether it fails for Big Sur, as well as for ARM.)

Though the errors are slightly different between them:

* https://build.macports.org/builders/ports-12_x86_64-builder/builds/68505/steps/install-port/logs/stdio
* https://build.macports.org/builders/ports-13_x86_64-builder/builds/32040/steps/install-port/logs/stdio

**Comment by mascguy** [comment]
Crap, it's also failing for ARM, albeit during the dependency installation phase. Hopefully this is easy to fix, as I thought (?) we had ARM support for some of our Java ports:

```
----> Installing dependency (17 of 78) 'bazel' with variants ''
--->  Fetching distfiles for bazel
DEBUG: java-portgroup: Trying to find JVM version: 11
DEBUG: No value for java JAVA_HOME was automatically discovered
DEBUG: Adding dependency on JDK fallback openjdk11-zulu
Error: bazel requires Java 11 but no such installation could be found.
Error: Failed to fetch bazel: missing required Java version
```

https://build.macports.org/builders/ports-13_arm64-builder/builds/21046/steps/install-dependencies/logs/stdio

**Comment by catap** [comment]
So, someone had broke java for arm?

**Comment by mascguy** [comment]
> So, someone had broke java for arm?

Hopefully not, this might simply be a bug/oversight in the Java portgroup. Need to investigate further.

**Comment by catap** [comment]
```
catap@Kirills-mini-m1 ~ % sudo port install openjdk11-zulu
--->  Fetching archive for openjdk11-zulu
--->  Attempting to fetch openjdk11-zulu-11.64.19_0.darwin_22.arm64.tbz2 from https://packages.macports.org/openjdk11-zulu
--->  Attempting to fetch openjdk11-zulu-11.64.19_0.darwin_22.arm64.tbz2 from https://nue.de.packages.macports.org/openjdk11-zulu
--->  Attempting to fetch openjdk11-zulu-11.64.19_0.darwin_22.arm64.tbz2 from https://fra.de.packages.macports.org/openjdk11-zulu
--->  Fetching distfiles for openjdk11-zulu
--->  Attempting to fetch zulu11.64.19-ca-jdk11.0.19-macosx_aarch64.tar.gz from https://cdn.azul.com/zulu/bin/
--->  Verifying checksums for openjdk11-zulu                                         
--->  Extracting openjdk11-zulu
--->  Configuring openjdk11-zulu
--->  Building openjdk11-zulu
--->  Staging openjdk11-zulu into destroot
Warning: openjdk11-zulu installs files outside the common directory structure.
--->  Installing openjdk11-zulu @11.64.19_0
```
Seems that openjdk11-zulu wasn't build by Buildbot...

**Comment by mascguy** [comment]
> The builds are failing very early in the process, for Monterey and Ventura Intel. (Still waiting to see whether it fails for Big Sur, as well as for ARM.)

Details for Monterey (paths trimmed to be relative, for readability):

```
Repository rule llvm_configure defined at:
  py310-tensorflow/work/bazel_build/149bd4da7eceb9ef1c48eadfd6b093af/external/llvm-raw/utils/bazel/configure.bzl:169:33: in <toplevel>
Analyzing: target //tensorflow/tools/pip_package:build_pip_package (53 packages loaded, 12 targets configured)
ERROR: An error occurred during the fetch of repository 'llvm-project':
   Traceback (most recent call last):
	File "py310-tensorflow/work/bazel_build/149bd4da7eceb9ef1c48eadfd6b093af/external/llvm-raw/utils/bazel/configure.bzl", line 146, column 25, in _llvm_configure_impl
		_overlay_directories(repository_ctx)
	File "py310-tensorflow/work/bazel_build/149bd4da7eceb9ef1c48eadfd6b093af/external/llvm-raw/utils/bazel/configure.bzl", line 64, column 13, in _overlay_directories
		fail(("Failed to execute overlay script: '{cmd}'\n" +
Error in fail: Failed to execute overlay script:
'/usr/bin/python3 py310-tensorflow/work/bazel_build/149bd4da7eceb9ef1c48eadfd6b093af/external/llvm-raw/utils/bazel/overlay_directories.py --src py310-tensorflow/work/bazel_build/149bd4da7eceb9ef1c48eadfd6b093af/external/llvm-raw --overlay py310-tensorflow/work/bazel_build/149bd4da7eceb9ef1c48eadfd6b093af/external/llvm-raw/utils/bazel/llvm-project-overlay --target .'
```

Details for Ventura (again, paths trimmed to relative):

```
Loading: 96 packages loaded
    currently loading: tensorflow ... (26 packages)
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/google/XNNPACK/archive/659147817805d17c7be2d60bd7bbca7e780f9c82.zip failed:
    class java.io.FileNotFoundException GET returned 404 Not Found

INFO: Repository llvm-project instantiated at:
  py310-tensorflow/work/tensorflow-2.12.0/WORKSPACE:15:14: in <toplevel>
  py310-tensorflow/work/tensorflow-2.12.0/tensorflow/workspace2.bzl:967:21: in workspace
  py310-tensorflow/work/tensorflow-2.12.0/tensorflow/workspace2.bzl:539:15: in _tf_repositories
  py310-tensorflow/work/tensorflow-2.12.0/third_party/llvm/setup.bzl:22:19: in llvm_setup
Repository rule llvm_configure defined at:
  py310-tensorflow/work/bazel_build/149bd4da7eceb9ef1c48eadfd6b093af/external/llvm-raw/utils/bazel/configure.bzl:169:33: in <toplevel>
Loading: 99 packages loaded
    currently loading: tensorflow ... (27 packages)
```

**Comment by mascguy** [comment]
> The builds are failing very early in the process, for Monterey and Ventura Intel. (Still waiting to see whether it fails for Big Sur, as well as for ARM.)

One bit of good news: So far, the Big Sur build is proceeding, without the failures seen for Monterey and Ventura. Hopefully it will succeed, keep your fingers crossed...

**Comment by mascguy** [comment]
> ```
> catap@Kirills-mini-m1 ~ % sudo port install openjdk11-zulu
> --->  Fetching archive for openjdk11-zulu
> --->  Attempting to fetch openjdk11-zulu-11.64.19_0.darwin_22.arm64.tbz2 from https://packages.macports.org/openjdk11-zulu
> --->  Attempting to fetch openjdk11-zulu-11.64.19_0.darwin_22.arm64.tbz2 from https://nue.de.packages.macports.org/openjdk11-zulu
> --->  Attempting to fetch openjdk11-zulu-11.64.19_0.darwin_22.arm64.tbz2 from https://fra.de.packages.macports.org/openjdk11-zulu
> --->  Fetching distfiles for openjdk11-zulu
> --->  Attempting to fetch zulu11.64.19-ca-jdk11.0.19-macosx_aarch64.tar.gz from https://cdn.azul.com/zulu/bin/
> --->  Verifying checksums for openjdk11-zulu                                         
> --->  Extracting openjdk11-zulu
> --->  Configuring openjdk11-zulu
> --->  Building openjdk11-zulu
> --->  Staging openjdk11-zulu into destroot
> Warning: openjdk11-zulu installs files outside the common directory structure.
> --->  Installing openjdk11-zulu @11.64.19_0
> ```
> 
> Seems that openjdk11-zulu wasn't build by Buildbot...

I think that was intentional, though perhaps not ideal. Theoretically that shouldn't cause the issue though; I'm guessing the Java PG isn't finding the installation.

**Comment by mascguy** [comment]
> I think that was intentional, though perhaps not ideal. Theoretically that shouldn't cause the issue though; I'm guessing the Java PG isn't finding the installation.

On further review, it looks like the Java PG is working correctly, and pulling in `openjdk11-zulu` correctly for ARM.

Instead, it appears that Bazel (the project, not our port logic) isn't finding the JDK. Still digging...

**Comment by catap** [comment]
BTW I quite supprised that bazel doesn't build JDK from scratch. Seem that it something which they need to work on the next release :)

**Comment by mascguy** [comment]
> The builds are failing very early in the process, for Monterey and Ventura Intel. (Still waiting to see whether it fails for Big Sur, as well as for ARM.)

Hmmmm... well, on the 2nd time around, the build for Monterey Intel didn't fail like it did before. So perhaps some upstream server was down, when files were being pulled? (Something in the realm of `https://storage.googleapis.com/mirror.tensorflow.org/github.com/google/...`.)

So there's hope that the Ventura Intel build might do the same. Fingers crossed...

**Comment by essandess** [comment]
That’s pretty impressive you all got this and the `bazel` port working. Thank you!

**Comment by mascguy** [comment]
Update: The builds for Bug Sur and Monterey almost complete successfully, but they fail at the very end with:

```
Execution platform: locul execution_config platform//:platform

2023-05-08 09:51:12.901685: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available cpu instructions in performance-critical of To enable the following instructions: SSE3 55E4.1 55E4.2, In other operations, rebuild tensorflow with the appropriate compiler flags.

Traceback (most recent call last):

File "py39 tensorflow/work/bazel build/e65bb5c129a76359f8df5ada6c4c0686/exerroot/org tensorflow/bazel-out/darwin-opt-exec-504E0418/
bin/tensorflow/create_tensorflow.python_api_t python_api_gen_v2.runfiles/org tensorflow/tensorflow/python/tools/api/generator/create_python api.py", line 22, in <module>
  From tensorflow.python.tools.api.generator import doc sres
File "py39-tensorflow/work/bazel build/e65bb5c129a76359f8d+Sade6c4c0686/execroot/org tensorflow/bazel-out/darwin opt exec-S04E0418/
bin/tensorflow/create tensorflow.python api tf python_api_gen_v2.runfiles/org tensorflow/tensorflow/python/init.py", line 37, in <module> File "py39-tensorflow/work/bazel build/e65bb5c129a76359f8df5da6c4c0686/execroot/org/tensorflow/bazel-out/darwin-opt-exer-504F0418/
from tensorflow.python.eager Import context
bin/tensorflow/create_tensorflow.python_api_t python_api_gen_v2.runfiles/org tensorflow/tensorflow/python/eager/context.py", line 20, In modules
from tensorflow.core.framework Import function_pb2 File "py39-tensorflow/work/bazel_build/e65bb5c129a76359fdf5ada6c4c0686/execroot/org tensorflow/bazel-out/daruin-opt-exec-504F0418/ bin/tensorflow/create_tensorflow.python_api tf python_api_gen_v2.runfiles/org tensorflow/tensorflow/core/framework/function_pb2.py", line 5, in codule> from google.protobuf. internal Import builder as builder

ImportError: cannot import name "builder from 'google.protobut.internal

(/opt/local/Library/Frameworks/Python, framework/Versions/3.9/lib/python3.5/site-packages/google/protobuf/internal/__init__.py)

Target //tensorflow/tools/pip_package:build_pip package failed to build

ERROR: py39-tensorflow/work/tensorflow-2.12.0/tensorflow/python/tools/BUILD:314:10 Middleman middlemen/tensorflow Spython Stools Ssaved Umodel ucl runfiles failed: (Exit 1): bash
```

**Comment by catap** [comment]
@mascguy I found the way to overstep that issue and it seems like insanity: https://stackoverflow.com/a/72494013

**Comment by mascguy** [comment]
> @mascguy I found the way to overstep that issue and it seems like insanity: https://stackoverflow.com/a/72494013

Yeah, I found that article too. Very interesting!

During your research, did you see any definitive info, re: Which version of `protobuf` is required now? I ask, as upstream is in the 4., territory, but we don't support that yet...

**Comment by cjones051073** [comment]
> @mascguy I found the way to overstep that issue and it seems like insanity: https://stackoverflow.com/a/72494013

Insanity is part of the course as far as I am concerned when it comes to Basel based builds...

**Comment by mascguy** [comment]
> > @mascguy I found the way to overstep that issue and it seems like insanity: https://stackoverflow.com/a/72494013
> 
> Insanity is part of the course as far as I am concerned when it comes to Bazel based builds...

After taking an initial look at the documentation, and getting a better understanding of what it provides, etc, it seems like an incredibly powerful build tool! But with the caveat that such functionality also requires a lot of learning.

I'm not saying it's perfect by any means - nor am I advocating that we expand it's use - but it does bring a lot to the table...

**Comment by catap** [comment]
> After taking an initial look at the documentation, and getting a better understanding of what it provides, etc, it seems like an incredibly powerful build tool! But with the caveat that the amount of functionality also requires a lot of learning.

I think that it is a kind of mistake to see bazel like a build tools. It is more like Docker or Nix which builds an application from different pieces.

It can be used to build an application, but it is more general things.
 
> I'm not saying it's perfect by any means - nor am I advocating that we expand its use - but it does bring a lot to the table...

The issue that it brings the table with it. The specific table which has more than one a kind of vendor lock-in to google.

**Pull request merged**
